---
title: "BINF 6309 Module 05: Variant Calling"
author: "Courtney Wong"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
bibliography: citations/bibliography.ris
---

# Overview

In this module, we walk through scripts that perform variant calling using the "DeepVariant" program [@Poplin], which uses Trimmomatic [@Bolger] and bwa-mem [@Li]. Additionally, SAMtools [@Li2] is used in this pipeline. The sample sequencing dataset used in this walkthrough is the same as the one used in Supernat et al to compare three different variant calling methods [@Supernat].

# Methods

## 1. Retrieve the reference human genome

The reference genome (human, release 27) is first obtained in `getGenome.sh`, shown below. This involves using `wget`, a network retriever. The output, specified by the `-O` flag, is saved locally as `GRCh38_reference.fa.gz`.

```{bash, code=readLines("getGenome.sh"), echo=TRUE, eval=FALSE}
```

## 2. Retrieve the sequencing data

The NGS reads are obtained in `getReads.sh` using `fastq-dump` to access NCBI's Sequence Read Archive. The accession number SRR6808334 is used to specify the reference. The `--split-files` flag specifies to separate Mate 1 (R1) from Mate 2 (R2) paired-end reads.

```{bash, code=readLines("getReads.sh"), echo=TRUE, eval=FALSE}
```

## 3. Quality trim the reads

Trimmomatic is used to remove poor-quality reads from the paired-end FASTQ files. A few notes on bioinformatics-related option flags used:

-  `PE`: indicates the data came from a paired-end run
-  `-phred33`: indicates to use the Phred+33 quality encoding method for the quality score
-  `HEADCROP`: specifies the number of bases to remove from the beginning of each read regardless of quality, set to 0
-  `ILLUMINACLIP`: specifies a path to TruSeq index sequences, in order to distinguish these manually-added adaptors during library preparation PCR from biological data
-  `LEADING` and `TRAILING`: specifies the minimum threshold for trimming the start and end of reads, set to 20
-  `MINLEN`: specifies the minimum read length for a read to be kept, set to 36

```{bash, code=readLines("trimReads.sh"), echo=TRUE, eval=FALSE}
```

## 4. Index the genome

The Burrows-Wheeler Alignment Tool is used to index the reference genome obtained in Step 1, to expedite the following alignment step. The BWT-SW algorithm is specified using the `-a` flag, and the input FASTA is specified with as an unflagged argument. Since no prefix is specified, the output database will have the same name as the input.

```{bash, code=readLines("indexGenome.sh"), echo=TRUE, eval=FALSE}
```

## 5. Align the reads

The trimmed reads obtained in Step 3 are aligned to the reference genome obtained in Step 1 in `alignReads.sh`. A few notes on bioinformatics-related option flags used:

-  `-R` specifies the read group header line, separated by tabs, in the SAM file that will be outputted
-  `-p` specifies the reference genome followed by the mate 1 and mate 2 FASTQ files

```{bash, code=readLines("alignReads.sh"), echo=TRUE, eval=FALSE}
```

## 6. Sort the aligned reads and save as BAM

The SAM files generated by Step 5 are sorted and converted into BAM files in `sort.sh` by using `samtools`. A few notes on option flags used:

-  `-@ 8` specifies 8 cores to be used in this multi-threaded program
-  `-m 4G` specifies 4 gigabytes of memory to be used in each of the 8 cores
-  `SRR6808334.sam` contains the aligned reads and will be sorted
-  `-o SRR6808334.bam` specifies where the sorted alignment will be saved, in BAM format

```{bash, code=readLines("sort.sh"), echo=TRUE, eval=FALSE}
```

## 7. Index the reads

The sorted BAM file from Step 6 is indexed in `indexReads.sh` using samtools once again. The `-b` flag indicates to use a BAI index.

```{bash, code=readLines("indexReads.sh"), echo=TRUE, eval=FALSE}
```

## 8. Identify variants!

This step, `runDeepVariant.sh`, is the heart of the pipeline and key part of this module. This script runs `DeepVariant` on the output from Step 7 to produce a VCF file, the format that stores information on variants.

First, error handling behavior of the script is specified in the line `set -euo pipefail`, which makes the program quit on the first encountered error, rather than proceeding and passing the error to the next part of the pipeline.

The next few lines define constants for paths, filenames, and version numbers. Following this, a directory structure is created using the defined paths and the `mkdir` command. Then, Linux packages are updated.

The block of code following the `if` statement ensures that `docker`, a platform commonly used in bioinformatics to run software in an isolated, neatly-packaged environment called a "container" and will be used in the following lines to run DeepVariant, is installed locally.

The next few lines ensure all the files are in the expected locations by copying over files generated in the previous steps, including the sorted BAM (Step 6), indexed BAM (Step 7), and indexed reference genome (Step 1 and 4).

One last line before running DeepVariant is called to pull the docker image, allowing DeepVariant to run in a dockerized container.

Finally, DeepVariant is run via Docker with the following key arguments:

-  `--model_type=WGS` specifies that the sample reads were obtained using whole-genome sequencing. This is one possible option out of the following: `[WGS,WES,PACBIO,ONT_R104,HYBRID_PACBIO_ILLUMINA]`
-  `--ref` specifies the path to the reference genome file, `GRCh38_reference.fa.gz`
-  `--reads` specifies the path to the NGS read file, which we have already sorted and saved as `SRR6808334.bam.sorted`
-  ` --output_vcf` and `--output_gvcf` specify where to save the variant calling outputs. Note that a GVCF stores information on all sites, whether a variant was detected or not, whereas a VCF stores only information on sites that contain variants.
-  `--num_shards` specifies the number of processors to use to run this program. This was set to 64 at the beginning of the script.

The final output of this pipeline is `SRR6808334.output.vcf.gz` and `SRR6808334.output.vcf.gz`.

# References